### PR TITLE
Online DDL docs: advanced usage

### PR DESCRIPTION
**Pull request by shlomi-noach**
Offering a page that illustrates how to achieve:

- Duplicate migration detection
- Duplicate migration indication
- Near instant REVERTs

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Affected (new) page preview: https://deploy-preview-944--vitess.netlify.app/docs/13.0/user-guides/schema-changes/advanced-usage/#near-instant-reverts

**Review by askdba — COMMENTED** [review]
LGTM

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+Two migrations sharing the same context and DDL are considered duplicate, and only one will run to completion.
+
+Consider this DDL:
+
+```sql
+alter table `customer` add column `status` int unsigned not null;
+```
+
+You can only run this change successfully once. Once it it applied, the column `status` exists; any attempt to run the migration again yields with error.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
You can only run this change successfully once. Once it it applied, the column `status` exists; any attempt to run the migration again yields an error.
```

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+Sometimes it is desireable to be able to retry a migration. For example, if you apply a migration on a sharded keyspace, where one or more of the shards can be down. In such scenario some shards receive and apply the DDL, while other shards do not, and are not away of its existence. Attempting to re-apply the same DDL will generate errors on the shards that have received and applied it on the first attempt.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
Sometimes it is desirable to be able to retry a migration. For example, if you apply a migration on a sharded keyspace, where one or more of the shards can be down. In such scenario some shards receive and apply the DDL, while other shards do not, and are not aware of its existence. Attempting to re-apply the same DDL will generate errors on the shards that have received and applied it on the first attempt.
```

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+`vtctl ApplySchema` accepts a `-request_context` flag. By default, Vitess auto-generates a unique context per execution of `vtctl ApplySchema`. You may supply your own value, which can be an arbitrary text (limited to `1024` characters). You may search for migrations with a particular context via `SHOW VITESS_MIGRATIONS LIKE '<context-value>'`. Also, any `SHOW VITESS_MIGRATIONS ...` command outputs the context value in the `migration_context` column.
+
+When Vitess meets a migration which has exact same DDL and exact same (non-empty) context as some older migration, it considers it as a _duplicate_. The new migration does get a `UUID` of its own, and tracked as a new migration. But if the previous migration (or, if there are multiple past duplicate migrations with same DDL and context, _any one of those_) is `complete`, then the new migration is also implicitly assumed to be `complete`.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
When Vitess meets a migration which has exact same DDL and exact same (non-empty) context as some older migration, it considers it as a _duplicate_. The new migration does get a `UUID` of its own, and is tracked as a new migration. But if the previous migration (or, if there are multiple past duplicate migrations with same DDL and context, _any one of those_) is `complete`, then the new migration is also implicitly assumed to be `complete`.
```

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+- As mentioned, the new migration will have its own `UUID`.
+- The new migration does get to be scheduled.
+- It will only be marked as `complete` if one previous identical migration (same DDL, same non-empty context) is likewise `complete`.
+- If, for example, there's a single previous identical migration which is `failed`, the new migration gets to be executed.
+- Continuing the above example: if the new migration is successful, the previous migration remains in `failed` state.
+- The decision whether to run the migration or to mark it as implicitly `complete` only takes place when the migration is scheduled to run.
+
+## Duplicate migration indication
+
+You may go one step beyond [duplicate migration detection](#duplicate-migration-detection) by explicitly supply migration UUIDs such that duplicate migrations are never submitted in the first place.
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
You may go one step beyond [duplicate migration detection](#duplicate-migration-detection) by explicitly supplying migration UUIDs such that duplicate migrations are never submitted in the first place.
```

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+Consider the following example, not `-uuid_list` flag:
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
Consider the following example, note `-uuid_list` flag:
```

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+Normally Vitess generates a `UUID` for each migration, thus having a new, unique ID per migration. With `-uuid_list`, you can force Vitess into using your own supplied UUID. There cannot be two migrations with the same `UUID`. Therefore, any subsequent submission of a migration with an already existing `UUID` is implicitly discarded. The 2nd call does return the migration `UUID`, but is internally discarded.
+
+This feature is useful for external management systems (e.g. schema deployment tools) which may want to control the identity of migrations. The external systems are thus able to own the UUID and can re-submit at will.
+
+Notes:
+
+- `-uuid_list` accepts zero or more comma separated UUID values
+- If empty, Vitess calculates UUID for the migrations
+- Number of supplied UUIDs must match the number of DDL statements in `-sql`
+- Each UUID must be in this similar format: `73380089_7764_11ec_a656_0a43f95f28a3`
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]: What is the acceptable format? Maybe some more examples or a regular expression will be helpful.

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+## Near instant REVERTs
+
+A [REVERT](../revertible-migrations/) operation is available for `CREATE`, `DROP` and `ALTER` statements. Both `CREATE` and `DROP` reverts are near instantaneous by design. An `ALTER`'s revert is fast, but not instantaneous. While it does not need to copy any table data, hence not proportional to table size and migration runtime, it does need to apply any changes made to the table since migration completion. This means tracking and applying binary log events, and the operation runtime is generally proportional to the time elapsed since migration completion.
+
+It is possible to preemptively initiate an "open-ended" revert, such that a new workflow prepares the grounds for a revert, but requires a user interaction to actually cut over.
+
+The use case and workflow is as follows:
+
+- A long migration runs. The user suspects there might be a risk to the schema change
+- As soon as the migration completes, the user issues an open ended revert, preparing the grounds to undoing the schema change
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
- As soon as the migration completes, the user issues an open ended revert, preparing the ground for undoing the schema change
```

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+As soon as the migration completes, we run:
+
+```sh
+$ vtctlclient ApplySchema -skip_preflight -ddl_strategy='online -postpone-completion -allow-concurrent' -sql "revert vitess_migration '29231906_776f_11ec_a656_0a43f95f28a3'" commerce
+3cc4ae0e_776f_11ec_a656_0a43f95f28a3
+```
+
+The above begins a `REVERT` migration that is open-ended (does not complete), via `-postpone-completion`. We also request the migration to run concurrently via `-allow-concurrent` such that it does not block any further "normal" migrations on other tables. Note `3cc4ae0e_776f_11ec_a656_0a43f95f28a3` is the UUID for the reverted migration.
+
+Finally, are we satisfied that the `drop index` migration went well? We issue:
```
Comment (deepthi) [inline content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]:
```suggestion
Finally, if we are satisfied that the `drop index` migration went well, we issue:
```

**Review by deepthi — APPROVED** [review]
💯 
Just a few spelling/grammar suggestions.  You may merge once they are addressed.

File: content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md (hunk 6 above)
Comment (shlomi-noach) [reply content/en/docs/13.0/user-guides/schema-changes/advanced-usage.md]: Clarified via link to RFC, with two clear examples.

**Comment by shlomi-noach** [comment]
All comments addressed.

**Pull request merged**